server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -160,4 +384,474 @@ private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request)
         }
         return request;
     }
+
+    private static boolean isPartOfPIT(SearchRequest request, ShardSearchContextId contextId) {
+        final PointInTimeBuilder pointInTimeBuilder = request.pointInTimeBuilder();
+        if (pointInTimeBuilder != null) {
+            return request.pointInTimeBuilder().getSearchContextId(null).contains(contextId);
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    protected void doRun(Map<SearchShardIterator, Integer> shardIndexMap) {
+        AbstractSearchAsyncAction.doCheckNoMissingShards(getName(), request, shardsIts, AbstractSearchAsyncAction::makeMissingShardsError);
+        final Map<CanMatchPreFilterSearchPhase.SendingTarget, NodeQueryRequest> perNodeQueries = new HashMap<>();
+        final String localNodeId = searchTransportService.transportService().getLocalNode().getId();
+        final int numberOfShardsTotal = shardsIts.size();
+        for (int i = 0; i < numberOfShardsTotal; i++) {
+            final SearchShardIterator shardRoutings = shardsIts.get(i);
+            assert shardRoutings.skip() == false;
+            assert shardIndexMap.containsKey(shardRoutings);
+            int shardIndex = shardIndexMap.get(shardRoutings);
+            final SearchShardTarget routing = shardRoutings.nextOrNull();
+            if (routing == null) {
+                failOnUnavailable(shardIndex, shardRoutings);
+            } else {
+                final String nodeId = routing.getNodeId();
+                // local requests don't need batching as there's no network latency
+                if (this.batchQueryPhase && localNodeId.equals(nodeId) == false) {
```

Comment:
high-level, I find that the batchQueryPhase check should be outside of the loop perhaps. If we want to disable bacthing entirely, we would like to do so more globally? What is the reason that we check it for every shard? is the value going to change while we are iterating?

Ideally, I think it would be great to have entirely the previous execution when batching is disabled? Would that just be a matter of delegating to super.doRun entirely or is there more to it?

Maybe what I am suggesting is not mutually exclusive with what you did: you still need to support a mixed mode based on nodes transport versions, but I think that would be only for the case where batching is enabled, and it falls back to not batching against nodes that don't support the new node level action?

Reply:
Ah right, this is just a leftover from the times before some other cleanups when I wasn't able to reuse the abstract parent action. Added the shortcut to super now and removed the conditional here.